ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ACSReader.scala

```diff
@@ -36,398 +33,129 @@ class FilterTableACSReader(
-    initialTasks.isEmpty
-  ) Source.empty
-  else {
-    val (signalQueue, signalSource) = Source
-      .queue[Unit](initialTasks.size)
-      .preMaterialize()(materializer)
-
-    val queueState = new QueueState(signalQueue, initialTasks)
-
-    signalSource
-      // in theory signals could be executed parallel without maintaining order with mapAsyncUnordered,
-      // but in practice with mapAsync we force the parallel processing to "wait" for a slow work item,
-      // which prevents excessive memory consumption in later stages.
-      .mapAsync(workerParallelism) { _ =>
-        val task = queueState.startTask()
-        work(task).map { case (result, nextTask) =>
-          queueState.finishTask(nextTask)
-          task -> result
-        }(ExecutionContext.parasitic)
-      }
-  }
-
-  /** Helper class to capture stateful  operations of pullWorkerSource
-    */
-  class QueueState[TASK: Ordering](
-      signalQueue: BoundedSourceQueue[Unit],
-      initialTasks: Iterable[TASK],
-  )(implicit errorLogger: ContextualizedErrorLogger) {
-    private val priorityQueue =
-      new mutable.PriorityQueue[TASK]()(implicitly[Ordering[TASK]].reverse)
-    private var runningTasks: Int = 0
-
-    initialTasks.foreach(addTask)
-
-    def startTask(): TASK = synchronized {
-      runningTasks += 1
-      priorityQueue.dequeue()
-    }
-
-    def finishTask(nextTask: Option[TASK]): Unit = synchronized {
-      nextTask match {
-        case None if priorityQueue.isEmpty && runningTasks == 1 =>
-          signalQueue.complete()
-
-        case newTask =>
-          runningTasks -= 1
-          newTask.foreach(addTask)
-      }
-    }
-
-    private def addTask(task: TASK): Unit = {
-      priorityQueue.enqueue(task)
-      signalQueue.offer(()) match {
-        case QueueOfferResult.Enqueued => ()
-        case QueueOfferResult.Dropped =>
-          throw ParticipantBackpressure
-            .Rejection(
-              "Cannot enqueue signal: dropped. ACS reader queue bufferSize not big enough."
+  def idSource(
+      pageBufferSize: Int
+  )(getPage: Long => Future[Vector[Long]]): Source[Long, NotUsed] = {
+    assert(pageBufferSize > 0)
+    Source
+      .unfoldAsync(0L) { fromExclusive =>
+        getPage(fromExclusive).map {
+          case empty if empty.isEmpty => None
+          case nonEmpty =>
+            Some(
+              nonEmpty.last -> nonEmpty
             )
-            .asGrpcError
-        case QueueOfferResult.Failure(f) =>
-          throw LedgerApiErrors.InternalError
-            .Buffer("Failed to enqueue in ACS reader queue state: Internal failure", Some(f))
-            .asGrpcError
-        case QueueOfferResult.QueueClosed =>
-          throw LedgerApiErrors.InternalError
-            .Buffer("Failed to enqueue in ACS reader queue state: Queue closed", None)
-            .asGrpcError
+        }(scala.concurrent.ExecutionContext.parasitic)
       }
-    }
+      .buffer(pageBufferSize, OverflowStrategy.backpressure)
```

Comment:
The default is to buffer 1 id page, right? Does the buffer size affect throughput?

Reply:
It can in theory if we need to wait for the next ID page for the merge sort to continue, hence the buffer 1